src/shims/native_lib/mod.rs

```diff
@@ -276,6 +278,147 @@ trait EvalContextExtPriv<'tcx>: crate::MiriInterpCxExt<'tcx> {
+                        // stripping provenance from them, since it'll later be
+                        // exposed recursively.
+                        let ptr = mplace.ptr();
+                        // Make sure the provenance of this allocation is exposed;
+                        // there must be one for this mplace to be valid at all.
+                        // The interpreter-level allocation itself is exposed in
+                        // visit_reachable_allocs.
+                        this.expose_provenance(ptr.provenance.unwrap())?;
+                        // Then get the actual bytes.
+                        let id = this
+                            .alloc_id_from_addr(
+                                ptr.addr().bytes(),
+                                mplace.layout.size.bytes_usize().try_into().unwrap(),
+                                /* only_exposed_allocations */ true,
+                            )
+                            .unwrap();
+                        let ptr_raw = this.get_alloc_bytes_unchecked_raw(id)?;
+                        // SAFETY: We know for sure that at ptr_raw the next layout.size bytes
+                        // are part of this allocation and initialised. They might be marked as
+                        // uninit in Miri, but all bytes returned by the isolated allocator are
+                        // initialised.
+                        unsafe {
+                            std::slice::from_raw_parts(ptr_raw, mplace.layout.size.bytes_usize())
+                                .to_vec()
+                                .into_boxed_slice()
+                        }
+                    }
+                    either::Either::Right(imm) => {
+                        // For immediates, we know the backing scalar is going to be 128 bits,
+                        // so we can just copy that.
+                        // TODO: is it possible for this to be a scalar pair?
+                        let scalar = imm.to_scalar();
+                        if scalar.size().bytes() > 0 {
+                            let bits = scalar.to_bits(scalar.size())?;
+                            bits.to_ne_bytes().to_vec().into_boxed_slice()
+                        } else {
+                            throw_ub_format!("attempting to pass a ZST over FFI: {}", imm.layout.ty)
+                        }
+                    }
+                };
+
+                ffi::CArg::Struct(strukt, bytes)
+            }
+            _ => throw_unsup_format!("unsupported argument type for native call: {}", v.layout.ty),
+        })
+    }
+
+    /// Parses an ADT to construct the matching libffi type.
+    fn adt_to_ffitype(
+        &self,
+        orig_ty: Ty<'_>,
+        adt_def: ty::AdtDef<'tcx>,
+        args: &'tcx ty::List<ty::GenericArg<'tcx>>,
+    ) -> InterpResult<'tcx, FfiType> {
+        // TODO: is this correct? Maybe `repr(transparent)` when the inner field
+        // is itself `repr(c)` is ok?
```

Comment:
Yes, that would be okay. But let's leave it to a future PR, since a `repr(transparent)` wrapper around e.g. `i32` should also be supported and that'll be a bit more interesting -- it's going to be an `Immediate`.